### PR TITLE
Simplify build on macOS

### PR DESCRIPTION
**Pull request by 0xced**
* Check for OpenSSL (and pkg-config) in the check_prereqs() function
* Simplify OpenSSL installation instructions
* Automatically pick the OpenSSL version installed by Homebrew by exporting the proper PKG_CONFIG_PATH environment variable
* Improve the error message if cmake can't find OpenSSL

The previous instructions were asking the user to add symbolic links inside  /usr/local/lib/ and /usr/local/lib/pkgconfig/ for OpenSSL related files. There is no need for a complicated setup with symbolic links when the PKG_CONFIG_PATH environment variable points to the right pkgconfig, which is now done in build-commons.sh.

### CODE REVIEW

File: eng/native/build-commons.sh
Comment (jkotas) [inline eng/native/build-commons.sh:68]:
```suggestion
            # It's important to _export_ it since build-commons.sh is sourced by other scripts such as build-native.sh
```

**Review by akoeplinger — APPROVED** [review]
This was bugging me for a long time, thanks for fixing it!

File: eng/pipelines/common/global-build-job.yml
Comment (akoeplinger) [inline eng/pipelines/common/global-build-job.yml]: looks like this needs to be `parameters.osGroup` and I think we can remove the `setScriptToEchoAndFailOnNonZero` to make it consistent with the other places.
```suggestion
          sh $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }}
```

File: eng/pipelines/libraries/build-job.yml
Comment (akoeplinger) [inline eng/pipelines/libraries/build-job.yml:74]: same here

```suggestion
          - script: sh $(Build.SourcesDirectory)/eng/install-native-dependencies.sh ${{ parameters.osGroup }}
```

File: eng/pipelines/common/global-build-job.yml
Comment (0xced) [reply eng/pipelines/common/global-build-job.yml]: I force-pushed 1956d5e1177e149b9d904f0c2175ae92fead7c42. But why is it different from the coreclr job? 🤯
https://github.com/dotnet/runtime/blob/89b2f425fadecd00fde47ad6a397e9065711ed61/eng/pipelines/coreclr/templates/build-job.yml#L107

File: eng/pipelines/libraries/build-job.yml
Comment (0xced) [reply eng/pipelines/libraries/build-job.yml]: I force-pushed 1956d5e1177e149b9d904f0c2175ae92fead7c42.

File: eng/pipelines/common/global-build-job.yml
Comment (akoeplinger) [reply eng/pipelines/common/global-build-job.yml]: I think that's because `osGroup` is part of a `variables` YML section there which is set to `${{ parameters.osGroup }}`.

**Comment by akoeplinger** [comment]
Looks like this gets further now but `runtime-live-build` and `runtime (Installer Build and Test coreclr OSX_x64 Debug)` seem to not run the install-native-dependencies.sh script so they fail right now.

**Comment by 0xced** [comment]
I tried to fix the `runtime` pipeline wit 8f3eeb50692a092af7e10071084ac04ab0d95cb5 but I'm not sure if it's the best way.

The `runtime-live-build` pipeline fails with this error and I have no idea why and no idea how to diagnose this. 😕
> /Users/runner/runners/2.166.4/work/1/s/eng/install-native-dependencies.sh: No such file or directory

File: eng/pipelines/installer/jobs/base-job.yml
Comment (safern) [inline eng/pipelines/installer/jobs/base-job.yml:435]: you don't need the `sh` do you?

**Comment by safern** [comment]
FYI: @bartonjs @wfurt

File: docs/workflow/requirements/macos-requirements.md
Comment (jashook) [inline docs/workflow/requirements/macos-requirements.md:28]: Does newer work? I may have incorrectly thought we were stuck on 1.1. /cc @wfurt @janvorli

File: eng/native/build-commons.sh
Comment (safern) [inline eng/native/build-commons.sh:71]: Should we check for other prereqs like `ICU`?

Comment (akoeplinger) [reply eng/native/build-commons.sh:71]: ICU won't be a system-wide prereq on some platforms like iOS though.

**Review by jashook — APPROVED** [review]
Minus the open question about openssl 1.1 lgtm

File: docs/workflow/requirements/macos-requirements.md
Comment (bartonjs) [reply docs/workflow/requirements/macos-requirements.md:28]: We used to be stuck on 1.0.2; now we work anything 1.0.1-1.1.1 (though only 1.1.1 is still in support)

Comment (safern) [reply docs/workflow/requirements/macos-requirements.md:28]: Yes it works. We no longer do dynamic linking. We just use it to build and we link at runtime against whatever is installed on the running machine. cc: @bartonjs

File: eng/native/build-commons.sh
Comment (safern) [reply eng/native/build-commons.sh:71]: That's true... these messages are useful to the user though, maybe detect if targetOS is not iOS/Android/tvOS?

File: docs/workflow/requirements/macos-requirements.md
Comment (wfurt) [inline docs/workflow/requirements/macos-requirements.md:35]: Do you know if one can run tests locally without the links? (we would not see it the in CI as tests do run on different machine)

File: src/libraries/Native/Unix/System.Security.Cryptography.Native/CMakeLists.txt
Comment (bartonjs) [inline src/libraries/Native/Unix/System.Security.Cryptography.Native/CMakeLists.txt:17]: Not all distros use the same packaging; so the Linux name should still hedge.  And the macOS one might also be different from uh... MacPorts?  Whatever the Homebrew competitor is.

```suggestion
    message(FATAL_ERROR "!!! Cannot find libssl and System.Security.Cryptography.Native cannot build without it. Try installing libssl-dev (on Linux, but this may vary by distro) or openssl (on macOS) !!!. See the requirements document for your specific operating system: https://github.com/dotnet/runtime/tree/master/docs/workflow/requirements.")
```

**Comment by wfurt** [comment]
Since you touching this, can you accommodate https://github.com/dotnet/runtime/issues/34150 ? Perhaps have `brew` as the primary option.

File: eng/native/build-commons.sh
Comment (bartonjs) [inline eng/native/build-commons.sh:69]: Presumably this should set the value additively, not as a replacement...

File: docs/workflow/requirements/macos-requirements.md
Comment (bartonjs) [reply docs/workflow/requirements/macos-requirements.md:35]: Assuming Homebrew is still not linking the libraries into /usr/local/lib/, we need to keep the runtime dependencies section.

But as long as src\libraries\Native\build-native.sh is sourcing the script that adds the default homebrew path to the pkg-config probe path, we can remove the pkgconfig-specific steps from the docs.

File: docs/workflow/requirements/macos-requirements.md
Comment (0xced) [reply docs/workflow/requirements/macos-requirements.md]: I removed the *or newer* in f0f57416acc1cf78d2e96f2ed50492d1cb719e06. Let's be conservative about version 1.1.

File: eng/pipelines/installer/jobs/base-job.yml
Comment (wfurt) [reply eng/pipelines/installer/jobs/base-job.yml:435]: *sh would be useful if you need particular shell flavor of if nstall-native-dependencies.sh  does not have executable bit.

Comment (0xced) [reply eng/pipelines/installer/jobs/base-job.yml]: Addressed in 277c5b5ba66c44a0862a9a536e6aec32df747de9.

File: eng/native/build-commons.sh
Comment (0xced) [reply eng/native/build-commons.sh:69]: I think replacement is correct here. `build-commons.sh` is sourced by other scripts and we really just need openssl. Also, the default directory will be searched anyway, see the pkg-config documentation about the `PKG_CONFIG_PATH` environment variable:
> PKG_CONFIG_PATH
>  A colon-separated  (on  Windows,  semicolon-separated)  list  of
>  directories to search for .pc files.  The default directory will
>  always be searched after searching  the  path;  the  default  is
>  libdir/pkgconfig:datadir/pkgconfig  where  libdir  is the libdir
>  for pkg-config and datadir is the datadir for pkg-config when it
>  was installed.

**Comment by 0xced** [comment]
> Since you touching this, can you accommodate #34150 ? Perhaps have brew as the primary option.

Addressed in b2530b11ce057e03e92010a98fb4579823b2a31b. I removed the manual installation instructions altogether, the ICU installation a few lines below only has instruction with Homebrew anyway.

File: eng/pipelines/common/global-build-job.yml
Comment (safern) [inline eng/pipelines/common/global-build-job.yml:30]: Merge with the if statement just below?

File: eng/pipelines/installer/jobs/base-job.yml
Comment (safern) [inline eng/pipelines/installer/jobs/base-job.yml:434]: Merge with the if statement just below?

File: eng/pipelines/libraries/build-job.yml
Comment (safern) [inline eng/pipelines/libraries/build-job.yml:73]: Merge with the if statement for checking disk usage?

File: eng/pipelines/common/global-build-job.yml
Comment (0xced) [reply eng/pipelines/common/global-build-job.yml:30]: Not sure it's a good idea, the conditions are the same but they serve different purposes. Also, *Install Build Dependencies* and *Disk Usage before Build* are already long step names.

Comment (safern) [reply eng/pipelines/common/global-build-job.yml:30]: What I mean is just use the same condition. Two different steps. Yaml is indentation based so you can have:

```yml
- ${{ if something }}:
  - script: command
     displayName: first command
  - script: command2
     displayName: second command

- script: command3
   displayName: third command
```

So in this example command and command2 are conditioned to something but command3 is unconditional. So what I mean is to move this step after the disk usage step and share the same if, not merge the two commands in one step.

Comment (0xced) [reply eng/pipelines/common/global-build-job.yml:30]: Got it, done in 43a3d87b96323df90d67326f252eb632c56611ab.

File: eng/pipelines/installer/jobs/base-job.yml
Comment (0xced) [reply eng/pipelines/installer/jobs/base-job.yml:434]: Done in 43a3d87b96323df90d67326f252eb632c56611ab.

File: eng/pipelines/libraries/build-job.yml
Comment (0xced) [reply eng/pipelines/libraries/build-job.yml]: Done in Done in 43a3d87b96323df90d67326f252eb632c56611ab.

File: eng/native/build-commons.sh
Comment (akoeplinger) [reply eng/native/build-commons.sh:71]: I don't think it's worth it, for ICU and other dependencies the usual cmake logic and error handling is straightforward.

File: eng/pipelines/libraries/build-job.yml
Comment (safern) [inline eng/pipelines/libraries/build-job.yml:78]: Nit: empty line in between tasks, it is easier to read the yml.

File: eng/pipelines/installer/jobs/base-job.yml
Comment (safern) [inline eng/pipelines/installer/jobs/base-job.yml:436]: Nit, empty line for readability.

File: eng/pipelines/common/global-build-job.yml
Comment (safern) [inline eng/pipelines/common/global-build-job.yml:32]: Nit, empty line for readability.

File: eng/native/build-commons.sh
Comment (safern) [reply eng/native/build-commons.sh:71]: Makes sense.

File: docs/workflow/requirements/macos-requirements.md
Comment (wfurt) [reply docs/workflow/requirements/macos-requirements.md]: If this is meant for developers, I'm wondering if we can add rpath so runtime would find them as well automatically. Or add note that somebody can set LD_LIBRARY_PATH as alternative to linking. 
In either case I would be nice if one does not have to do any linking and direct file manipulation.

**Review by wfurt — APPROVED** [review]
LGTM. thanks.

File: eng/pipelines/libraries/build-job.yml
Comment (0xced) [reply eng/pipelines/libraries/build-job.yml:78]: Fixed in c6f19d622e7a713fdb1a2b84fc2ae45137704cb5.

File: eng/pipelines/installer/jobs/base-job.yml
Comment (0xced) [reply eng/pipelines/installer/jobs/base-job.yml:436]: Fixed in c6f19d622e7a713fdb1a2b84fc2ae45137704cb5.

File: eng/pipelines/common/global-build-job.yml
Comment (0xced) [reply eng/pipelines/common/global-build-job.yml:32]: Fixed in c6f19d622e7a713fdb1a2b84fc2ae45137704cb5.

File: eng/pipelines/common/global-build-job.yml
Comment (safern) [inline eng/pipelines/common/global-build-job.yml:31]: I think you accidentally brought back, `sh` into these steps.

Comment (0xced) [reply eng/pipelines/common/global-build-job.yml:31]: It was not accidental, see 7037603edce1a0d008685d209a9422beeacdf4ce which reverted the change. All the build were failing in Azure pipelines due to the removal of calling through `sh`.

Comment (safern) [reply eng/pipelines/common/global-build-job.yml:31]: We have never needed the usage of sh, could you point me to the build error?

Comment (0xced) [reply eng/pipelines/common/global-build-job.yml:31]: I can't find the corresponding Azure build but here's the error I observed:
> /Users/runner/runners/2.166.4/work/_temp/56691a53-a761-472c-b9ab-a9b4eb004f5d.sh: line 1: /Users/runner/runners/2.166.4/work/1/s/eng/install-native-dependencies.sh: Permission denied

All the jobs running the  `install-native-dependencies.sh` script without running through `sh` would fail with this error. This is why I reverted it in 7037603edce1a0d008685d209a9422beeacdf4ce.

Note that I wrote the scripts to run through `sh` because all other `install-native-dependencies.sh` were run through `sh`, you can observe them on the master branch.

**Comment by 0xced** [comment]
Anything holding the merge?

File: eng/pipelines/common/global-build-job.yml
Comment (wfurt) [reply eng/pipelines/common/global-build-job.yml:31]:
```
wfurt-runtime furt$ ls -al eng/install-native-dependencies.sh
-rw-r--r--  1 furt  staff  1293 Apr 28 18:03 eng/install-native-dependencies.sh
```

it seems like that scrips does not have executable bit as it should @safern 
Perhaps something we should fix independently. (or lump here)

Comment (safern) [reply eng/pipelines/common/global-build-job.yml:31]: Would you like to fix this in a follow up PR?

Comment (0xced) [reply eng/pipelines/common/global-build-job.yml:31]: @safern I'll leave it up to you.

Comment (safern) [reply eng/pipelines/common/global-build-job.yml:31]: Ok, it would be nice to fix it (just to be consistent across all places where we invoke bash scripts in the yml), so if you have time that would be great.

**Comment by EgorBo** [comment]
Oops, it broke android build (when we build it on macOS)

**Pull request merged**
